### ISSUE TITLE
Spanification

### ISSUE THREAD

**Issue by cocowalla** [dazinator/DotNet.Glob#48]
With the recent 'spanification' of .NET Core, in particular the addition of the Span-based `FileSystemEnumerable` and `System.IO.Path` Span-based methods, I'm wondering if it would be possible to add Span-based methods to `Glob`? The point would be to avoid allocating strings when performing Glob-based matching.

**Comment by dazinator** [comment]
Interesting idea. I'd have to do some research on this. Always open to any PRs that improve perf.

**Comment by dazinator** [comment]
Few things that spring to mind.

DotNet.Glob uses the string length in a few places to shortcut processing. Suppose you ask it to match a glob pattern that has a min length requirement of 5 - e.g `*12345*` and you match this glob pattern against a few thousand strings. It won't bother evaluating any of those strings that are less than 5 in length - it can immediately reject them because they fail the min length heuristics. If we add support for span then need to consider if we lose this length informatioin and the ability to do this short circuiting. Could that lead to a perf loss in some circumstances?

**Comment by cocowalla** [comment]
Not sure what you mean about losing the length information?

```
ReadOnlySpan<char> mySpan = "myergen".AsSpan();
var len = mySpan.Length; <--- = 7
```

**Comment by cocowalla** [comment]
A typical use for globbing is matching paths in a file system, where you'd recurse down through a directory and call `Glob.IsMatch` on each path to see if it's a match. Until recently, that meant allocating rather a lot of strings - but with the new `Span`-based `System.IO.Path` APIs, it's actually possible to do this with very little allocations, which means that callers may have a _span_, rather than a _string_. I was thinking of overloading Glob.IsMatch:

```
public bool IsMatch(string subject)
public bool IsMatch(ReadOnlySpan<char> subject)
```

There would be quite a bit of work to get that working behind the scenes though.

**Comment by dazinator** [comment]
@cocowalla  thanks for the detailed explaination - this makes sense. It sounds like this would be a good future feature to add when things are a bit more stable then. Let me know when you think that point is reached and we can re-look at this.

**Comment by dazinator** [comment]
Some related info here: https://blogs.msdn.microsoft.com/jeremykuhne/2018/03/09/custom-directory-enumeration-in-net-core-2-1/

**Comment by pwiens** [comment]
I added span support and an additional target framework. #62 
Once netstandard2.1 ships netcoreapp2.1 can be changed. @dazinator I'm actually using this library with FileSystemEnumerable right now haven't tested the perf different with Span<T> but I suspect it helps.

**Comment by dazinator** [comment]
Thanks @pwiens - have merged the PR, can close this one now ;-)

**Comment by dazinator** [comment]
I am going to do a little bit more work on this.

**Comment by dazinator** [comment]
I have:

1. Added a test that uses `ReadOnlySpan<char>` on .NET Core 2.1.
2. Added `.NET Standard 2.1` as a compilation target, in addition to `.NET Core 2.1` to that as other platforms implement .net standard they will also be able to consume this new API. 
3. Adjusted the compilation / preprocessor directives a bit to my liking!
4.. Removed the overload at the evaluator level, one less virtual method in the call chain should be better.
5. Updated the README.

I'll release version `3.0.0` shortly.

**Comment by dazinator** [comment]
This has been released in version `3.0.0` - thanks for contributing!

**Comment by pwiens** [comment]
Thanks man! First meaningful contribution to open source after 11 years in the industry. :)
